ldclient_scoped.go

```diff
@@ -0,0 +1,621 @@
+package ldclient
+
+import (
+	gocontext "context"
+	"sync"
+
+	"github.com/launchdarkly/go-sdk-common/v3/ldcontext"
+	"github.com/launchdarkly/go-sdk-common/v3/ldmigration"
+	"github.com/launchdarkly/go-sdk-common/v3/ldreason"
+	"github.com/launchdarkly/go-sdk-common/v3/ldvalue"
+	ldevents "github.com/launchdarkly/go-sdk-events/v3"
+	"github.com/launchdarkly/go-server-sdk/v7/interfaces"
+)
+
+// LDScopedClient is a wrapper around LDClient that lets you specify the
+// evaluation context to be used for all operations, rather than taking the
+// evaluation context as a parameter every time you call a method.
+//
+// A LDScopedClient is created by calling [LDClient.ForContext]. This sets the
+// initial context to be used for all operations:
+//
+//	userContext := ldcontext.New("user-key")
+//	scopedClient := client.ForContext(userContext)
+//	scopedClient.BoolVariation("flag-key", false)
+//
+// Scoped contexts are mutable, to facilitate incrementally building up a
+// multi-context representing the current scope. You should create a new scoped
+// client for each logical scope where the contexts are different. For instance,
+// if you are using a scoped client in an HTTP request handler, you should create
+// a new scoped client for each request. You should only use the mutable features
+// of LDScopedClient to add new contexts, not to change or remove existing data.
+// A scoped client is thread-safe, so you can safely use it from multiple
+// goroutines.
+//
+// You may add new contexts with [LDScopedClient.AddContext]. The scoped client's
+// contexts so far are combined into a multi-context whenever the scoped client
+// is used. This is useful when more contexts become available later in the
+// lifecycle of a request. For instance, you might have a `user` context that is
+// available early in the request, but you also want to evaluate flags with a
+// `company` context that is only available later:
+//
+//	company := fetchCompanyForUser(user)
+//	scopedClient.AddContext(ldcontext.NewWithKind("company", company.Id))
+//	scopedClient.BoolVariation("enterprise-features", false)
+//
+// You can also overwrite a context that was previously added, by calling
+// [LDScopedClient.OverwriteContextByKind]. This can be used when an existing
+// context needs to be updated with new data:
```

Comment:
```suggestion
// You can also overwrite a context that you previously added, by calling
// [LDScopedClient.OverwriteContextByKind]. Use this when you need to update an existing
// context with new data:
```
optional suggestion, updates for active voice